scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,186 @@
+      * This may not be able to be pure PostgeSQL, as the `etag` is generated in Go, so that is an extra challenge
+    * Populate the table upon Index Report fetching
+      * If the entry does not exist in the table (the case for all Index Reports created prior to this release), then create the Index Report and populate the table
+      * This essentially forces a re-indexing of all previously indexed images, whether it is required or not, which may be costly
+
+### Random Deletions
+
+When it comes to variables outside "versioned scanners" there is no currently known clean way to handle the re-indexing.
+
+Something like creating our own "versioned scanner" to track other types of changes is not a solution:
+
+* Updating the version of the custom "versioned scanner" will only affect that scanner, and not the others which would all be skipped upon re-indexing
+* It is not feasible to track every single change in ClairCore which may affect Index Reports
+
+Instead, we will randomly delete Index Reports from the ClairCore database.
+Doing this solves both the remaining aspects of the re-indexing problem and the database ever-growing problem,
+as unneeded Index Reports are permanently deleted and needed reports may just be regenerated.
+
+**Note**: this approach also solves the "versioned scanner" update scenario; however, we opt to implement the [`checkManifest`](#reuse-claircores-checkmanifest-step)
+step for faster turnaround upon "versioned scanner" updates.
+
+Of course, determining a timeframe for deletion is non-trivial. Deleting too quickly means Index Reports may be corrected
+faster, but too many (probably) unnecessary requests to the image registry and (probably) unnecessary CPU/memory/disk usage.
+Deleting too slowly means resource usage will be much less, but it may take (too) long to correct the Index Report.
+Similarly, we would need to ensure not all Index Reports are all re-indexed at the same time (hence random deletions)
+to ensure spikes are minimized.
+
+The deletions will be chosen from a range between **one week** and **one month**. We believe this is a sufficient timeframe
+which balances speed and resource consumption.
+
+This will all be tracked in a new table (see [Manifest Metadata Table](#manifest-metadata-table) below).
+
+The table will be queried periodically for passed timestamps. These rows will be deleted and the related Index Reports will be deleted from ClairCore's tables.
+
+#### Alternative - Client Delete Requests
+
+Scanner V4 clients (Central and Sensor) may inform the Indexer of an image deletion and request the Index Report deleted
+
+However:
+
+* It is non-trivial to pinpoint all places where images are deleted in Central and Sensor
+* The proposed method is simpler, as it is all done from within the Indexer
+
+### Manifest Metadata Table
+
+The new Manifest Metadata table will look like the following:
+
+| manifest_id   | expiration                    |
```

Comment:
Does the `expiration` column indicate the time at which we are required to forcefully recalculate the manifest_id for the versioned scanner?

Reply:
A little bit of background:

Claircore indexes manifests (we call them images, but you'll find before we pass the image to Claircore to index, we convert it into a `claircore.Manifest`). These manifests are stored in Claircore's Indexer DB alongside the Index Report. The hash ID we use for the manifest if the same ID as the Index Report.

This column indicates when we will declare the manifest obsolete and should be garbage collected. We will run a separate garbage collector to check this table for expired manifests, and then it will delete the manifests from this table as well as Claircore's Indexer table. This process of deleting the manifest from Claircore will delete the related Index Report as well.